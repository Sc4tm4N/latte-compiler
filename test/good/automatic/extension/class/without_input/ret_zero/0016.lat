class a {

  void print() {
    printString("a.print");
    a();
  }

  int a() {
    printString("a.a");
    return 0;
  }

  int ref() {
    printString("a.ref");
    return 0;
  }
}

class b extends a {

  void print() {
    printString("b.print");
    a();
  }

  int a() {
    printString("b.a");
    return b();
  }

  int b() {
    printString("b.b");
    return ref();
  }
}

class c extends b {

  void print() {
    printString("c.print");
    a();
  }

  int a() {
    printString("c.a");
    return b() * self.c();
  }

  int c() {
    printString("c.c");
    return 2 + c();
  }

  int ref() {
    printString("c.ref");
    return 666;
  }
}

int c() {
  return 20;
}

int main() {
  a a = new a;
  a.print();

  printString("----");

  b b = new b;
  b.print();

  printString("----");

  c c = new c;
  c.print();

  printString("----");

  a ab = new b;
  ab.print();

  printString("----");

  a ac = new c;
  ac.print();

  printString("----");

  b bc = new c;
  bc.print();

  return 0;
}
