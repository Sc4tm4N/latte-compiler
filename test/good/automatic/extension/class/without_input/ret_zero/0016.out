a.print
a.a
----
b.print
b.a
b.b
a.ref
----
c.print
c.a
b.b
c.ref
c.c
----
b.print
b.a
b.b
a.ref
----
c.print
c.a
b.b
c.ref
c.c
----
c.print
c.a
b.b
c.ref
c.c
